src/main/java/com/coreos/jetcd/internal/impl/WatchImpl.java

```diff
@@ -48,234 +73,275 @@
-            events.removeIf((e) -> e.getKv().getModRevision() <= lastRevision);
-          }
-          watcher.setLastRevision(
-              watchResponse
-                  .getEvents(watchResponse.getEventsCount() - 1)
-                  .getKv().getModRevision());
-
-          if (watcher.callback != null) {
-            watcher.callback.onWatch(
-                Util.toHeader(watchResponse.getHeader(), watchResponse.getCompactRevision()),
-                Util.toEvents(events));
-          }
-        } else {
-          watcher.setLastRevision(watchResponse.getHeader().getRevision());
-        }
-      }
+  private synchronized void processWatchResponse(WatchResponse watchResponse) {
+    // prevents grpc on sending watchResponse to a closed watch client.
+    if (this.isClosed()) {
+      return;
+    }
+
+    if (watchResponse.getCreated()) {
+      processCreate(watchResponse);
+    } else if (watchResponse.getCanceled()) {
+      processCanceled(watchResponse);
     } else {
-      // if the watcher is not canceling, cancel it.
-      CompletableFuture<Boolean> completableFuture = this.pendingCancelFutures
-          .get(watchResponse.getWatchId());
-      if (this.pendingCancelFutures.putIfAbsent(watcher.getWatchID(), completableFuture) == null) {
-        cancelWatch(watchResponse.getWatchId());
-      }
+      processEvents(watchResponse);
     }
   }
 
-  /**
-   * resume all the watchers.
-   */
-  private void resumeWatchers(WatcherImpl[] watchers) {
-    for (WatcherImpl watcher : watchers) {
-      if (watcher.callback != null) {
-        watcher.callback.onResuming();
-      }
-      watch(watcher.getKey(), getResumeWatchOptionWithWatcher(watcher), watcher.callback);
+  private synchronized void processError(Throwable t) {
+    // prevents grpc on sending error to a closed watch client.
+    if (this.isClosed()) {
+      return;
+    }
+
+    Status status = Status.fromThrowable(t);
+    if (this.isHaltError(status) || this.isNoLeaderError(status)) {
+      this.notifyWatchers(EtcdExceptionFactory.newEtcdException(status.getDescription()));
+      this.closeGrpcWatchStreamObserver();
+      this.cancelSet.clear();
+      return;
+    }
+    // resume with a delay; avoiding immediate retry on a long connection downtime.
+    scheduledExecutorService.schedule(this::resume, 500, TimeUnit.MILLISECONDS);
+  }
+
+  private synchronized void resume() {
+    this.closeGrpcWatchStreamObserver();
+    this.cancelSet.clear();
+    this.resumeWatchers();
+  }
+
+  // closeGrpcWatchStreamObserver closes the underlying grpc watch stream.
+  private void closeGrpcWatchStreamObserver() {
+    if (this.grpcWatchStreamObserver == null) {
+      return;
+    }
+    this.grpcWatchStreamObserver.onCompleted();
+    this.grpcWatchStreamObserver = null;
+  }
+
+  private boolean isNoLeaderError(Status status) {
+    if (status.getDescription() == null) {
+      return false;
+    }
+    return status.getDescription().contains("no leader");
```

Comment:
I can build a map of etcd server error to jetcd error. then use that map to determine "no leader" error. I can do that in a separate pr.